packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -866,14 +880,75 @@ export default function makeKernelKeeper(kernelStorage, kernelSlog) {
     kvStore.set(`${kernelSlot}.data.slots`, capdata.slots.join(','));
   }
 
-  function cleanupAfterTerminatedVat(vatID) {
+  function removeVat(vatID) {
+    // Delete primary swingstore records for this vat, in preparation
+    // for (slow) deletion. After this, swingstore exports will omit
+    // this vat. This is called from the kernel's terminateVat, which
+    // initiates (but does not complete) deletion.
+    snapStore.stopUsingLastSnapshot(vatID);
+    transcriptStore.stopUsingTranscript(vatID);
+  }
+
+  /**
+   * Perform some cleanup work for a specific (terminated but not
+   * fully-deleted) vat, possibly limited by a budget. Returns 'done'
+   * (where false means "please call me again", and true means "you
+   * can delete the vatID now"), and a count of how much work was done
+   * (so the runPolicy can decide when to stop).
+   *
+   * @param {string} vatID
+   * @param {number} [budget]
+   * @returns {{ done: boolean, cleanups: number }}
+   *
+   */
+  function cleanupAfterTerminatedVat(vatID, budget = undefined) {
     insistVatID(vatID);
+    let cleanups = 0;
+    const work = {
+      exports: 0,
+      imports: 0,
+      kv: 0,
+      snapshots: 0,
+      transcripts: 0,
+    };
+    let spend = _did => false; // returns "stop now"
+    if (budget !== undefined) {
+      assert.typeof(budget, 'number');
+      spend = (did = 1) => {
+        assert(budget !== undefined); // hush TSC
+        cleanups += did;
+        budget -= did;
+        return budget <= 0;
+      };
+    }
+    const logWork = () => {
+      const w = {};
+      for (const prop of Object.getOwnPropertyNames(work)) {
+        if (work[prop] !== 0) {
+          w[prop] = work[prop];
+        }
+      }
+      kernelSlog?.write({ type: 'vat-cleanup', vatID, work: w });
+    };
+
+    // TODO: it would be slightly cheaper to walk all kvStore keys in
+    // order, and act on each one according to its category (c-list
+    // export, c-list import, vatstore, other), so we use a single
+    // enumeratePrefixedKeys() call each time. Until we do that, the
+    // last phase of the cleanup (where we've deleted all the exports
+    // and imports, and are working on the remaining keys) will waste
+    // two DB queries on each call. OTOH, those queries will probably
+    // hit the same SQLite index page as the successful one, so it
+    // probably won't cause any extra disk IO. So we can defer this
+    // optimization for a while. Note: when we implement it, be
+    // prepared to encounter the clist entries in eiher order (kref
+    // first or vref first), and delete the other one in the same
+    // call, so we don't wind up with half an entry.
+
```

Comment:
I think it's fine to run the three queries and keep the specialized processing steps distinct in their own loops.
```suggestion
```

Reply:
Yeah, and it has the benefit that cleanup is broken up into very distinctive sections (first exports, then imports, then the rest of the kvstore, then snapshots, then finally transcripts).

Walking the kvStore in lexical order, and reacting to each key according to its value, would yield a somewhat-random intermixing of exports and imports, because we hit the `v7.c.koNN` section before we reach the `v7.c.o-NN` or `v7.c.o+NN` sections. 

Given that the koids were mostly allocated by a series of PushPrice actions, it's conceivable that we'll have a couple exports, then a couple imports, etc. And given that we only delete 5 c-list entries at a time before triggering a BOYD, it's possible that one cleanup run will see mostly exports, then the next might see mostly imports. So the cost of each cleanup run might vary considerably, which seems like a negative.

It might cost more DB queries (like three per cleanup run, which is negligible in the grand scheme of things), but having distinct phases would certainly be easier to talk about. "We're in the exports phase, we should switch to the imports phase in four days", with each phase yielding very consistent costs.